### DISCUSSION TITLE
How to customize a ListItem widget?

### DISCUSSION THREAD

**Discussion by Tim-Cao** [Textualize/textual#3994] [Q&A] [answered]
```python
from typing import Optional

from textual.app import App, ComposeResult
from textual.widgets import *
from textual.widgets import *


class MenuItem(ListItem):
    def __init__(self, name: Optional[str] = None, id: Optional[str] = None) -> None:
        super().__init__()
        self.name = name
        self.id = id

    def compose(self) -> ComposeResult:
        yield Text(self.name, self.id)


class Text(Static):
    pass


class Demo(App[None]):
    def compose(self) -> ComposeResult:
        yield Header()
        yield ListView(MenuItem("Text Item", "nav-1"))
        yield Footer()


if __name__ == "__main__":
    Demo().run()
```

When do it in this way, I met this error. Could anyone help me? Thanks
![image](https://github.com/Textualize/textual/assets/52661397/2f87be67-4c87-4079-9de4-f21174fa3014)

**Comment by davep** [comment, marked as answer]
There's a couple of different problems with your code. First off, Textual widgets have a [`name`](https://textual.textualize.io/api/dom_node/#textual.dom.DOMNode.name) and an [`id`](https://textual.textualize.io/api/dom_node/#textual.dom.DOMNode.id) property anyway, so if you're overriding the `__init__` and you want to use those you'll need to pass them to the `super().__init__` call. Secondly, when you instantiate your `Text` widget, which inherits from `Static`, you're passing more positional arguments than [`Static`'s `__init__` is designed to take](https://textual.textualize.io/widgets/static/#textual.widgets.Static); it has only the one positional argument, the rest are keyword arguments.

A clean starting point for your code would be something like this:

```python
from textual.app import App, ComposeResult
from textual.widgets import Static, ListView, ListItem, Header, Footer


class Demo(App[None]):
    def compose(self) -> ComposeResult:
        yield Header()
        yield ListView(ListItem(Static("Text Item", id="nav-1")))
        yield Footer()


if __name__ == "__main__":
    Demo().run()
```

Assuming you want to add to what you were trying to do, but this was a basic example, a starting point more akin to what you were doing might be this:

```python
from textual.app import App, ComposeResult
from textual.widgets import Static, ListView, ListItem, Header, Footer


class Text(Static):
    pass


class MenuItem(ListItem):

    def __init__(self, text: str, id: str | None = None) -> None:
        super().__init__(id=id)
        self._text = text

    def compose(self) -> ComposeResult:
        yield Text(self._text, id=self.id)


class Demo(App[None]):

    def compose(self) -> ComposeResult:
        yield Header()
        yield ListView(MenuItem("Text Item", "nav-1"))
        yield Footer()


if __name__ == "__main__":
    Demo().run()
```

**Reply by Tim-Cao** [reply]
How to know which is positional argument?
![image](https://github.com/Textualize/textual/assets/52661397/500245c8-85b6-42f1-b8aa-2eb5cccb2ee6)

**Reply by davep** [reply]
Anything before the `*` in the above would be positional[^1]; anything after would be keyword-only. You can read more about this notation in the standard Python documentation.

[^1]: Technically optionally a keyword too but that's not important here.

**Reply by Tim-Cao** [reply]
I see. Thanks @davep
